sequencer/src/block2.rs

```diff
@@ -0,0 +1,144 @@
+// Question: jellyfish uses `ark_std` everywhere instead of `std`.
+// Is there a chance of badness if we (as a downstream user) use `std`?
+use std::mem::size_of;
+
+#[allow(dead_code)] // TODO temporary
+pub struct BlockPayload {
+    payload: Vec<u8>,
+    // tx_ranges: Vec<Range<u32>>, // not convinced we need this
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Tx {
+    // namespace: u32, // TODO how many bytes for namespace id?
+    payload: Vec<u8>,
+}
+
+impl BlockPayload {
+    #[allow(dead_code)] // TODO temporary
+    fn build(txs: impl IntoIterator<Item = Tx>) -> Self {
+        // tx_table[i] is the end index (exclusive) of the ith tx
+        // so that the payload bytes of the ith tx is
+        // tx_bodies[tx_table[i-1]..tx_table[i]].
+        // edge case: tx_table[-1] is defined as 0.
+        //
+        // TODO clarification for future docs:
+        // perhaps the final entry of tx_table is redundant:
+        // it should always equal the length of the flattened tx bytes.
+        // But this might not always be the case.
+        // Example: If the flattened tx bytes are not a monolith at the end
+        // of the payload then the final tx_table entry is needed.
+        let mut tx_table = Vec::new();
+
+        // concatenation of all tx payloads
+        let mut tx_bodies = Vec::new();
+
+        // build tx_table, tx_bodies
+        let mut end: u32 = 0;
+        for tx in txs.into_iter() {
+            // TODO idiomatic usize -> u32 conversion?
+            // Is panic acceptable? If not then we need to return `Result`.
```

Comment:
Yeah I think after @shenkeyao finishes the `BlockPayload` trait this function will return a `Result`

Reply:
I put a comment about this in the code. I'll make it return `Result` after I have an upstream trait.